endtoend/endtoend_test.go

```diff
@@ -296,3 +303,44 @@ func (r *testRunner) testBeaconChainSync(ctx context.Context, g *errgroup.Group,
 
 	return nil
 }
+
+func (r *testRunner) testDoppelGangerProtection(ctx context.Context) error {
+	// Exit if we are running from
+	// the previous release.
+	if r.config.UsePrysmShValidator {
+		return nil
+	}
+	g, ctx := errgroup.WithContext(ctx)
+	// Follow same parameters as older validators.
+	validatorNum := int(params.BeaconConfig().MinGenesisActiveValidatorCount)
+	beaconNodeNum := e2e.TestParams.BeaconNodeCount
+	if validatorNum%beaconNodeNum != 0 {
+		return errors.New("validator count is not easily divisible by beacon node count")
+	}
+	validatorsPerNode := validatorNum / beaconNodeNum
+	valIndex := beaconNodeNum + 1
+
+	// Replicate starting up validator clien 0 to test doppleganger protection.
+	valNode := components.NewValidatorNode(r.config, validatorsPerNode, valIndex, validatorsPerNode*0)
+	g.Go(func() error {
+		return valNode.Start(ctx)
+	})
+	if err := helpers.ComponentsStarted(ctx, []e2etypes.ComponentRunner{valNode}); err != nil {
+		return fmt.Errorf("validator not ready: %w", err)
+	}
+	logFile, err := os.Create(path.Join(e2e.TestParams.LogPath, fmt.Sprintf(e2e.ValidatorLogFileName, valIndex)))
+	if err != nil {
+		return fmt.Errorf("unable to open log file: %v", err)
+	}
+	r.t.Run("doppelganger found", func(t *testing.T) {
+		assert.NoError(t, helpers.WaitForTextInFile(logFile, "Could not succeed with doppelganger check"), "Failed to carry out doppelganger check")
```

Comment:
I'm a little puzzled by this line. It looks like you are expecting the text `Could not succeed with doppelganger check` in the log file, but that would mean the doppelganger check failed i.e. it was not able to determine whether a doppelganger is present. So it seems to me that this evaluation does not work at all.

Reply:
I looked at the code. Seems like doppelganger returns an error in the following conditions:
- Something goes wrong in preprocessing inputs
- Some issue in the RPC call to the beacon node to check doppelganger
- Nil response from beacon node so unsafe to proceed, return err
- Duplicate validator running, return err

It is safer to instead return `bool, error` to handle the different conditions. Otherwise, the return error for everything can be misleading. We should change the function signature of doppelganger in this PR perhaps

Reply:
So there is no need to do that, it is simpler just to check for duplicates existing in the logs. 
https://github.com/prysmaticlabs/prysm/blob/develop/validator/client/validator.go#L438

You would just need to check the logs for this
```
Duplicate instances exists in the network for validator keys
```